nfs/src/rpc_task.cpp

```diff
@@ -474,6 +488,412 @@ void rpc_task::run_setattr()
+
+    assert(!is_async());
+
+    // We should have completed all the reads before sending the response to caller.
+    assert(num_of_reads_issued_to_backend == 0);
+
+    if (status)
+    {
+        // Non-zero status indicates failure, reply with error in such cases.
+        reply_error(status);
+        return;
+    }
+
+    const size_t count = bytes_vector.size();
+
+    // Create an array of iovec struct
+    struct iovec iov[count];
+
+    // Fetch the caller requested size.
+    const size_t req_size = rpc_api.readfile_task.get_size();
+    size_t remaining_size = req_size;
+
+    for (size_t i = 0; (i < count && remaining_size > 0); i++)
+    {
+        if (remaining_size >= bytes_vector[i].length)
+        {
+            /*
+             * If the first chunk itself is empty, then there is no need to
+             * look further, so just send empty response as we reach here only
+             * in the case of success.
+             * TODO: Check what will happen if the length of inbetween vectors is 0.
+             */
+            if ((i==0) && (bytes_vector[i].length == 0))
+            {
+                reply_iov(nullptr, 0);
+                return;
+            }
+
+            iov[i].iov_base = (void*)bytes_vector[i].get_buffer();
+            iov[i].iov_len = bytes_vector[i].length;
+            remaining_size -= bytes_vector[i].length;
+        }
+        else
+        {
+            iov[i].iov_base = (void*)bytes_vector[i].get_buffer();
+            iov[i].iov_len = remaining_size;
+            remaining_size = 0;
+            break;
+        }
+    }
+
+    // Send response to caller.
+    reply_iov(iov, count);
+}
+
+struct readfile_context
+{
+    rpc_task *task;
+    struct bytes_chunk* bc;
+
+    readfile_context(
+        rpc_task *task_,
+        struct bytes_chunk* bc_):
+        task(task_),
+        bc(bc_)
+    {}
+};
+
+static void readfile_callback(
+    struct rpc_context* /* rpc */,
+    int rpc_status,
+    void *data,
+    void *private_data)
+{
+    struct readfile_context *ctx = (readfile_context*) private_data;
+
+    rpc_task *task = ctx->task;
+    assert (task->num_of_reads_issued_to_backend > 0);
+
+    struct bytes_chunk *bc = ctx->bc;
+    assert(bc != nullptr);
+
+    /*
+     * Release the lock that we held on the membuf since the data is now written to it.
+     * The lock is needed only to write the data and not to just read it.
+     * Hence it is safe to read this membuf even beyond this point.
+     */
+    bc->get_membuf()->clear_locked();
+
+    // Free the context.
+    delete ctx;
+
+    auto res = (READ3res*)data;
+
+    AZLogDebug("readfile_callback:: is_async:{} Bytes read: {} eof: {}, requested_bytes: {} off: {}",
+               task->is_async(),
+               res->READ3res_u.resok.count, res->READ3res_u.resok.eof,
+               bc->length,
+               bc->offset);
+
+    const int status = (task->succeeded(rpc_status, RSTATUS(res))) ? 0 : -nfsstat3_to_errno(RSTATUS(res));
```

Comment:
this is not entirely correct, since succeeded() may return false for rpc status failure and in that case nfs status might be success. I've added a new function status() for this.